httpreplay/cobweb.py

```diff
@@ -175,3 +180,245 @@ def handle(self, s, ts, protocol, sent, recv):
 
 class SmtpProtocol(Protocol):
     """Interprets the SMTP protocol."""
+
+    def init(self, *args, **kwargs):
+        self.request = SmtpRequest()
+        self.reply = SmtpReply()
+
+        # Current smtp command from client
+        self.command = None
+
+        # Last smtp server response code
+        self.rescode = 0
+
+        # Last smtp server response message
+        self.message = ""
+
+        # Used in handler() to determine if the TCP stream
+        # was closed or is finished
+        self.stream = None
+
+        # Contains the functions to be called when this command is the first
+        # string in a request message
+        self._commands = {
+            "ehlo": self.handle_hostname,
+            "helo": self.handle_hostname,
+            "mail": self.handle_mail,
+            "rcpt": self.handle_rcpt,
+            "auth": self.handle_auth,
+            "etrn": None, "turn": None, "atrn": None, "size": None,
+            "etrn": None, "pipelining": None, "chunking": None, "data":None,
+            "dsn": None, "rset": None, "vrfy": None, "help": None,
+            "quit": None, "noop": None, "expn": None, "binarymime": None,
+            "relay": None, "size": None, "starttls": None, "checkpoint": None,
+            "enhancedstatuscodes": None, "8bitmime": None, "send": None
+        }
+
+        # Contains the functions to be called
+        # in the request handler after the server sents this response code
+        self._res_codes = {
+            334: self.handle_auth_serv_response,
+            354: self.handle_mailbody
+        }
+
+    def handle(self, s, ts, protocol, sent, recv):
+        if protocol != "tcp":
+            self.parent.handle(s, ts, protocol, sent, recv)
+            return
+
+        if self.stream is None:
+            self.stream = self.parent.tcp.streams[s]
+
+        self.parse_request(sent)
+        self.parse_reply(recv)
+
+        if self.stream.state in ["conn_finish", "conn_closed"]:
+            self.parent.handle(s, ts, "smtp", self.request, self.reply)
+
+    def handle_hostname(self, data):
+        if len(data) > 1:
+            self.request.hostname = data[1]
+
+    def handle_rcpt(self, data):
+        if len(data) < 2:
+            return
+        for val in data:
+            self.request.mail_to.extend(re.findall('<(.*?)>', val, re.DOTALL))
+
+    def handle_mail(self, data):
+        if len(data) < 2:
+            return
+        for val in data:
+            self.request.mail_from.extend(re.findall('<(.*?)>', val, re.DOTALL))
+
+    def handle_mailbody(self, data):
+        headers_mes = data.split("\r\n\r", 1)
```

Comment:
Just double checking, is this supposed to be `\r\n\r` or should there be another trailing `\n`?